File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.py
from collections import defaultdict

def solution(topping):
    cnt = 0
    
    left_set = set()
    right_cnt = defaultdict(int)
    
    for t in topping:
        right_cnt[t] += 1
        
    for t in topping:
        left_set.add(t)
        right_cnt[t] -= 1
        
        if right_cnt[t] == 0:
            del right_cnt[t]
            
        if len(left_set) == len(right_cnt):
            cnt += 1

    return cnt

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.js
function solution(topping) {
    let cnt = 0;
    
    const leftSet = new Set();
    const rightObj = {};
    let rightCnt = 0;
    
    for (const t of topping) {
        if (!rightObj[t]) {
            rightObj[t] = 1;
            rightCnt += 1;
        } else {
            rightObj[t] += 1;
        }
    }
    
    for (const t of topping) {
        leftSet.add(t);
        rightObj[t] -= 1;
        
        if (rightObj[t] === 0) {
            delete rightObj[t];
            rightCnt -= 1;
        }
        
        if (leftSet.size === rightCnt) {
            cnt += 1;
        }
    }
    
    return cnt;
}